tests/unit/s2n_resume_test.c

```diff
@@ -1618,5 +1618,127 @@ int main(int argc, char **argv)
+         */
+        {
+            DEFER_CLEANUP(struct s2n_stuffer ticket_stuffer = { 0 }, s2n_stuffer_free);
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&ticket_stuffer, 0));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint8(&ticket_stuffer, S2N_STATE_WITH_SESSION_TICKET));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&ticket_stuffer, sizeof(server_state)));
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&ticket_stuffer, server_state, sizeof(server_state)));
+            EXPECT_SUCCESS(s2n_stuffer_write_bytes(&ticket_stuffer,
+                    tls12_ticket_with_ems, sizeof(tls12_ticket_with_ems)));
+
+            size_t ticket_size = s2n_stuffer_data_available(&ticket_stuffer);
+            uint8_t *ticket_bytes = s2n_stuffer_raw_read(&ticket_stuffer, ticket_size);
+            EXPECT_NOT_NULL(ticket_bytes);
+
+            DEFER_CLEANUP(struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            EXPECT_NOT_NULL(conn);
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+            /* Test that deserialize modifies the connection in limited ways.
+             *
+             * No mechanism exists to do more than a shallow comparison of two connections.
+             * To prove that a shallow comparison is sufficient, we need to prove
+             * that s2n_deserialize_resumption_state does not modify the memory
+             * associated with pointers on the connection. To prove that, we can
+             * test that s2n_deserialize_resumption_state can successfully operate
+             * on an s2n_connection with a limited number of its pointers initialized.
+             */
+            struct s2n_connection empty_conn = { 0 };
+            struct s2n_crypto_parameters crypto_params = { .cipher_suite = &s2n_null_cipher_suite };
+            empty_conn.secure = &crypto_params;
+            empty_conn.mode = S2N_CLIENT;
+            /* We can safely assume that a connection doesn't modify its config */
+            empty_conn.config = conn->config;
+            EXPECT_SUCCESS(s2n_connection_set_session(&empty_conn, ticket_bytes, ticket_size));
+            EXPECT_SUCCESS(s2n_free(&empty_conn.client_ticket));
+
+            /* Trigger the deserialization failure as late as possible.
+             * The last byte is optional, so drop the last two bytes.
+             */
+            ticket_size -= 2;
+
+            /* Test the connection is not modified by a failed deserialize */
+            uint8_t saved_conn[sizeof(struct s2n_connection)] = { 0 };
+            EXPECT_MEMCPY_SUCCESS(saved_conn, conn, sizeof(struct s2n_connection));
+            uint8_t saved_secure[sizeof(struct s2n_crypto_parameters)] = { 0 };
+            EXPECT_MEMCPY_SUCCESS(saved_secure, conn->secure, sizeof(struct s2n_crypto_parameters));
+            EXPECT_FAILURE_WITH_ERRNO(s2n_connection_set_session(conn, ticket_bytes, ticket_size),
```

Comment:
In the case that a failed `s2n_connection_set_session` modified the connection, wouldn't this test still pass since we are using the same ticket with the same parameters?

Reply:
edit: missed that this is happening on different connections.